File: packages/lib-shared-py/src/lib_shared/database/utils/api_data_loader/api_data_loader_models.py
from typing import Callable, Generic, Type, TypeVar

import attr

from lib_shared.database.utils.data_loader.data_loader import (
    DataLoaderModel,
    DataLoaderParams,
    TType,
)
from lib_shared.http.utils.constants import HttpContentType, HttpMethod

TResponse = TypeVar("TResponse")


@attr.s(auto_attribs=True, kw_only=True)
class ApiDataLoaderParams(DataLoaderParams[TType], Generic[TType, TResponse]):
    uri: str
    transformer: Callable[[TResponse], list[TType]]
    response: Type[TResponse] | None = None
    method: HttpMethod | None = HttpMethod.GET
    content_type: HttpContentType | None = HttpContentType.JSON
    headers: dict | None = {}
    params: dict | None = {}


class ApiDataLoaderModel(DataLoaderModel):
    def __init__(
        self,
        params: ApiDataLoaderParams,
    ) -> None: ...

File: packages/service-job-js/src/data/utils/ApiDataLoader/ApiDataLoader.models.ts
import { type SourcedEntityResourceModel } from '@lib/model/data/SourcedEntityResource/SourcedEntityResource.models';
import { type PartialArrayModel } from '@lib/shared/core/core.models';
import { type XmlNodeModel } from '@lib/shared/data/utils/parseXml/parseXml.models';
import { type HTTP_METHOD, type HTTP_RESPONSE_TYPE } from '@lib/shared/http/http.constants';
import {
  type DataLoaderModel,
  type DataLoaderParamsModel,
} from '@service/job/data/utils/DataLoader/DataLoader.models';

export type ApiDataLoaderParamsModel<
  TType extends SourcedEntityResourceModel,
  TResponse = unknown,
  TResponseType extends HTTP_RESPONSE_TYPE = HTTP_RESPONSE_TYPE.JSON,
> = DataLoaderParamsModel<TType> & {
  headers?: Record<string, unknown>;
  method?: HTTP_METHOD;
  params?: Record<string, unknown>;
  responseType?: TResponseType;
  uri: string;
  transformer(
    response: TResponseType extends HTTP_RESPONSE_TYPE.XML ? XmlNodeModel : TResponse,
  ): Promise<PartialArrayModel<TType>>;
};

export type ApiDataLoaderModel<TType extends SourcedEntityResourceModel> = DataLoaderModel<TType>;
